internal/cmdexport/orchestrators/tekton.go

```diff
@@ -0,0 +1,232 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orchestrators
+
+import (
+	"bytes"
+	"path"
+
+	"github.com/GoogleContainerTools/kpt/internal/cmdexport/types"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// TektonPipeline wraps a Pipeline object in Tekton.
+// @see https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md
+type TektonPipeline struct {
+	APIVersion string              `yaml:"apiVersion,omitempty"`
+	Kind       string              `yaml:",omitempty"`
+	Metadata   *TektonMetadata     `yaml:",omitempty"`
+	Spec       *TektonPipelineSpec `yaml:",omitempty"`
+	task       *TektonTask
+}
+
+func (p *TektonPipeline) Init(config *types.PipelineConfig) Pipeline {
+	p.APIVersion = "tekton.dev/v1beta1"
+	p.Kind = "Pipeline"
+	p.Metadata = &TektonMetadata{Name: "run-kpt-functions"}
+
+	workspaceName := "shared-workspace"
+	pipelineWorkspace := &TektonWorkspace{
+		Name: workspaceName,
+	}
+
+	taskName := "run-kpt-functions"
+
+	taskConfig := &TektonTaskConfig{PipelineConfig: config, Name: taskName}
+	task := new(TektonTask).Init(taskConfig)
+
+	pipelineTaskWorkspace := &TektonWorkspace{
+		Name:      "source",
+		Workspace: workspaceName,
+	}
+	pipelineTask := &TektonPipelineTask{
+		Name:       "kpt",
+		TaskRef:    &TektonPipelineTaskRef{Name: taskName},
+		Workspaces: []*TektonWorkspace{pipelineTaskWorkspace},
+	}
+
+	p.task = task
+	p.Spec = &TektonPipelineSpec{
+		Workspaces: []*TektonWorkspace{pipelineWorkspace},
+		Tasks:      []*TektonPipelineTask{pipelineTask},
+	}
+
+	return p
+}
+
+// Generate outputs a multi-doc yaml that contains Tekton Pipeline Object and a Tekton Task object.
+func (p *TektonPipeline) Generate() (out []byte, err error) {
+	task, err := p.task.Generate()
+
+	if err != nil {
+		return
+	}
+
+	pipeline, err := yaml.Marshal(p)
+
+	if err != nil {
+		return
+	}
+
+	// Concat the two yaml docs.
+	b := &bytes.Buffer{}
+
+	b.Write(task)
+	b.WriteString("---\n")
+	b.Write(pipeline)
+
+	return b.Bytes(), nil
+}
+
+// TektonPipelineSpec describes the spec of a Pipeline.
+type TektonPipelineSpec struct {
+	Workspaces []*TektonWorkspace    `yaml:",omitempty"`
+	Tasks      []*TektonPipelineTask `yaml:",omitempty"`
+}
+
+// // TektonPipelineTaskRef represents a task in a Tekton Pipeline object.
+type TektonPipelineTask struct {
+	Name       string                 `yaml:",omitempty"`
+	TaskRef    *TektonPipelineTaskRef `yaml:"taskRef,omitempty"`
+	RunAfter   []string               `yaml:"runAfter,omitempty"`
+	Workspaces []*TektonWorkspace     `yaml:",omitempty"`
+}
+
+// TektonPipelineTaskRef represents a taskRef field in a task of a Tekton Pipeline object.
+type TektonPipelineTaskRef struct {
+	Name string `yaml:",omitempty"`
+}
+
+// TektonMetadata contains metadata to describe a resource object.
+type TektonMetadata struct {
```

Comment:
Suggest to rename the types for pipeline to start with `Pipeline` and the types for tasks to start with `Task`.